File: Minimal/src/Brontide.Minimal.Host/ProviderArtifactAcquirer.fs
namespace Brontide.Minimal.Host

open System
open System.IO

[<StructuralEquality; StructuralComparison>]
type ProviderArtifactSourceId = private ProviderArtifactSourceId of string

[<RequireQualifiedAccess>]
module ProviderArtifactSourceId =
    let create value =
        if String.IsNullOrWhiteSpace value then invalidArg (nameof value) "A provider artifact source identity is required."
        ProviderArtifactSourceId value

    let value (ProviderArtifactSourceId value) = value

type ProviderArtifactAcquisitionFile =
    { RelativePath: string
      Sha256: string
      Length: int64 }

type ProviderArtifactAcquisitionRequest =
    { ExpectedSource: ProviderArtifactSourceId
      Identity: ProviderArtifactSetId
      Files: ProviderArtifactAcquisitionFile list
      ExecutablePath: string
      Arguments: string list
      MaxTotalBytes: int64 }

type IProviderArtifactSource =
    abstract Identity: ProviderArtifactSourceId
    abstract OpenRead: relativePath: string -> Stream option

type ProviderArtifactAcquisitionResult =
    { SourceIdentity: ProviderArtifactSourceId
      TransportCode: string
      PublisherEvidenceCode: string
      AdmissionCode: string
      Staged: StagedProviderArtifactSet option }
    member this.IsStaged = this.Staged.IsSome

type ProviderArtifactAcquirer(store: ContentAddressedProviderStore, transactionRoot: string) =
    let syncRoot = obj ()
    let root = Path.GetFullPath transactionRoot

    let safeRelativePath (path: string) =
        not (String.IsNullOrWhiteSpace path)
        && not (Path.IsPathFullyQualified path)
        && not (path.Contains '\\')
        && (path.Split '/' |> Array.forall (fun segment -> segment.Length > 0 && segment <> "." && segment <> ".."))

    let isDigest (value: string) =
        not (String.IsNullOrEmpty value)
        && value.Length = 64
        && value |> Seq.forall (fun character ->
            (character >= '0' && character <= '9') || (character >= 'A' && character <= 'F'))

    let isValid (request: ProviderArtifactAcquisitionRequest) =
        let structural =
            not (List.isEmpty request.Files)
            && request.MaxTotalBytes > 0L
            && not (request.Arguments |> List.exists String.IsNullOrEmpty)
            && safeRelativePath request.ExecutablePath
            && (request.Files |> List.forall (fun file ->
                safeRelativePath file.RelativePath && isDigest file.Sha256 && file.Length >= 0L))
            && (request.Files |> List.map _.RelativePath |> List.distinct |> List.length) = request.Files.Length
            && (request.Files |> List.exists (fun file -> file.RelativePath = request.ExecutablePath))
        if not structural then
            false
        else
            let withinLimit, _ =
                request.Files
                |> List.fold (fun (valid, total) file ->
                    if not valid || file.Length > request.MaxTotalBytes - total then false, total
                    else true, total + file.Length) (true, 0L)
            let files = request.Files |> List.map (fun file -> { RelativePath = file.RelativePath; Sha256 = file.Sha256 })
            withinLimit
            && ProviderArtifactSetIdentity.compute files request.ExecutablePath request.Arguments = request.Identity

    let refused source code =
        { SourceIdentity = source
          TransportCode = code
          PublisherEvidenceCode = "publisher-evidence-not-evaluated"
          AdmissionCode = "admission-not-attempted"
          Staged = None }

    let copyExact (input: Stream) (output: Stream) length =
        let buffer = Array.zeroCreate<byte> 81920
        let mutable remaining = length
        let mutable exact = true
        while exact && remaining > 0L do
            let read = input.Read(buffer, 0, min buffer.Length (int remaining))
            if read = 0 then exact <- false
            else
                output.Write(buffer, 0, read)
                remaining <- remaining - int64 read
        exact && input.ReadByte() = -1

    let transportException (error: exn) =
        error :? IOException
        || error :? UnauthorizedAccessException
        || error :? ObjectDisposedException
        || error :? NotSupportedException

    let combineRelative (basePath: string) (relative: string) =
        Path.GetFullPath(Path.Combine(basePath, relative.Replace('/', Path.DirectorySeparatorChar)))

    let deleteTree path =
        let rec remove attempt =
            if Directory.Exists path then
                try
                    Directory.EnumerateFiles(path, "*", SearchOption.AllDirectories)
                    |> Seq.iter (fun file -> File.SetAttributes(file, FileAttributes.Normal))
                    Directory.Delete(path, true)
                with
                | :? IOException when attempt < 4 ->
                    Threading.Thread.Sleep 25
                    remove (attempt + 1)
                | :? UnauthorizedAccessException when attempt < 4 ->
                    Threading.Thread.Sleep 25
                    remove (attempt + 1)
        remove 0

    let acquireMember (source: IProviderArtifactSource) transaction (memberFile: ProviderArtifactAcquisitionFile) =
        try
            match source.OpenRead memberFile.RelativePath with
            | None -> Some "acquisition-member-unavailable"
            | Some stream ->
                use input = stream
                let destination = combineRelative transaction memberFile.RelativePath
                match Path.GetDirectoryName destination |> Option.ofObj with
                | Some parent -> Directory.CreateDirectory parent |> ignore
                | None -> failwith "An acquired member must have a parent directory."
                use output = new FileStream(destination, FileMode.CreateNew, FileAccess.Write, FileShare.None)
                if copyExact input output memberFile.Length then None
                else Some "acquisition-length-mismatch"
        with error when transportException error ->
            Some "acquisition-transport-failed"

    do
        if isNull (box store) then nullArg (nameof store)
        if String.IsNullOrWhiteSpace transactionRoot then invalidArg (nameof transactionRoot) "A transaction root is required."
        Directory.CreateDirectory root |> ignore

    member _.Acquire(request: ProviderArtifactAcquisitionRequest, source: IProviderArtifactSource) =
        if isNull (box source) then nullArg (nameof source)
        lock syncRoot (fun () ->
            if not (isValid request) then
                refused request.ExpectedSource "acquisition-invalid"
            elif source.Identity <> request.ExpectedSource then
                refused request.ExpectedSource "acquisition-source-mismatch"
            else
                let transaction = Path.Combine(root, $".acquire-{Guid.NewGuid():N}")
                Directory.CreateDirectory transaction |> ignore
                try
                    let failure =
                        request.Files
                        |> List.sortBy _.RelativePath
                        |> List.tryPick (acquireMember source transaction)
                    match failure with
                    | Some code -> refused request.ExpectedSource code
                    | None ->
                        let declaration: ProviderArtifactSet =
                            { Identity = request.Identity
                              SourceRoot = transaction
                              Files = request.Files |> List.map (fun file ->
                                  { RelativePath = file.RelativePath; Sha256 = file.Sha256 })
                              ExecutablePath = request.ExecutablePath
                              Arguments = request.Arguments }
                        match store.Stage declaration with
                        | ProviderArtifactStagingResult.Staged staged ->
                            { SourceIdentity = request.ExpectedSource
                              TransportCode = "transport-completed"
                              PublisherEvidenceCode = "publisher-evidence-not-evaluated"
                              AdmissionCode = "staged"
                              Staged = Some staged }
                        | ProviderArtifactStagingResult.Refused failure ->
                            { SourceIdentity = request.ExpectedSource
                              TransportCode = "transport-completed"
                              PublisherEvidenceCode = "publisher-evidence-not-evaluated"
                              AdmissionCode = failure.Code
                              Staged = None }
                finally
                    deleteTree transaction)

File: Reference/src/Brontide.Reference.Studio/ProviderArtifactAcquirer.cs
namespace Brontide.Reference.Studio;

public readonly record struct ProviderArtifactSourceId
{
    private ProviderArtifactSourceId(string value) => Value = value;

    public string Value { get; }

    public static ProviderArtifactSourceId Create(string value)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(value);
        return new(value);
    }

    public override string ToString() => Value;
}

public sealed record ProviderArtifactAcquisitionFile(string RelativePath, string Sha256, long Length);

public sealed record ProviderArtifactAcquisitionRequest(
    ProviderArtifactSourceId ExpectedSource,
    ProviderArtifactSetId Identity,
    IReadOnlyList<ProviderArtifactAcquisitionFile> Files,
    string ExecutablePath,
    IReadOnlyList<string> Arguments,
    long MaxTotalBytes);

public interface IProviderArtifactSource
{
    ProviderArtifactSourceId Identity { get; }

    Stream? OpenRead(string relativePath);
}

public sealed record ProviderArtifactAcquisitionResult
{
    private ProviderArtifactAcquisitionResult(
        ProviderArtifactSourceId sourceIdentity,
        string transportCode,
        string admissionCode,
        StagedProviderArtifactSet? staged)
    {
        SourceIdentity = sourceIdentity;
        TransportCode = transportCode;
        AdmissionCode = admissionCode;
        Staged = staged;
    }

    public ProviderArtifactSourceId SourceIdentity { get; }

    public string TransportCode { get; }

    public string PublisherEvidenceCode => "publisher-evidence-not-evaluated";

    public string AdmissionCode { get; }

    public StagedProviderArtifactSet? Staged { get; }

    public bool IsStaged => Staged is not null;

    internal static ProviderArtifactAcquisitionResult TransportRefused(
        ProviderArtifactSourceId sourceIdentity,
        string code) => new(sourceIdentity, code, "admission-not-attempted", null);

    internal static ProviderArtifactAcquisitionResult Admitted(
        ProviderArtifactSourceId sourceIdentity,
        ProviderArtifactStagingResult staging) =>
        new(sourceIdentity, "transport-completed", staging.Code, staging.Staged);
}

public sealed class ProviderArtifactAcquirer
{
    private readonly object _sync = new();
    private readonly ContentAddressedProviderStore _store;
    private readonly string _transactionRoot;

    public ProviderArtifactAcquirer(ContentAddressedProviderStore store, string transactionRoot)
    {
        ArgumentNullException.ThrowIfNull(store);
        ArgumentException.ThrowIfNullOrWhiteSpace(transactionRoot);
        _store = store;
        _transactionRoot = Path.GetFullPath(transactionRoot);
        Directory.CreateDirectory(_transactionRoot);
    }

    public ProviderArtifactAcquisitionResult Acquire(
        ProviderArtifactAcquisitionRequest request,
        IProviderArtifactSource source)
    {
        ArgumentNullException.ThrowIfNull(request);
        ArgumentNullException.ThrowIfNull(source);
        lock (_sync)
        {
            if (request.Files is null || request.Arguments is null)
            {
                return ProviderArtifactAcquisitionResult.TransportRefused(
                    request.ExpectedSource,
                    "acquisition-invalid");
            }

            request = request with
            {
                Files = Array.AsReadOnly(request.Files.ToArray()),
                Arguments = Array.AsReadOnly(request.Arguments.ToArray()),
            };
            if (!IsValid(request))
            {
                return ProviderArtifactAcquisitionResult.TransportRefused(
                    request.ExpectedSource,
                    "acquisition-invalid");
            }

            if (source.Identity != request.ExpectedSource)
            {
                return ProviderArtifactAcquisitionResult.TransportRefused(
                    request.ExpectedSource,
                    "acquisition-source-mismatch");
            }

            var transaction = Path.Combine(_transactionRoot, $".acquire-{Guid.NewGuid():N}");
            Directory.CreateDirectory(transaction);
            try
            {
                foreach (var member in request.Files.OrderBy(file => file.RelativePath, StringComparer.Ordinal))
                {
                    Stream? input;
                    try
                    {
                        input = source.OpenRead(member.RelativePath);
                    }
                    catch (Exception exception) when (IsTransportException(exception))
                    {
                        return ProviderArtifactAcquisitionResult.TransportRefused(
                            request.ExpectedSource,
                            "acquisition-transport-failed");
                    }

                    if (input is null)
                    {
                        return ProviderArtifactAcquisitionResult.TransportRefused(
                            request.ExpectedSource,
                            "acquisition-member-unavailable");
                    }

                    using (input)
                    {
                        var destination = CombineRelative(transaction, member.RelativePath);
                        Directory.CreateDirectory(Path.GetDirectoryName(destination)!);
                        try
                        {
                            using var output = new FileStream(destination, FileMode.CreateNew, FileAccess.Write, FileShare.None);
                            if (!CopyExact(input, output, member.Length))
                            {
                                return ProviderArtifactAcquisitionResult.TransportRefused(
                                    request.ExpectedSource,
                                    "acquisition-length-mismatch");
                            }
                        }
                        catch (Exception exception) when (IsTransportException(exception))
                        {
                            return ProviderArtifactAcquisitionResult.TransportRefused(
                                request.ExpectedSource,
                                "acquisition-transport-failed");
                        }
                    }
                }

                var staging = _store.Stage(new ProviderArtifactSet(
                    request.Identity,
                    transaction,
                    request.Files.Select(file => new ProviderArtifactFile(file.RelativePath, file.Sha256)).ToArray(),
                    request.ExecutablePath,
                    request.Arguments));
                return ProviderArtifactAcquisitionResult.Admitted(request.ExpectedSource, staging);
            }
            catch (Exception exception) when (IsTransportException(exception))
            {
                return ProviderArtifactAcquisitionResult.TransportRefused(
                    request.ExpectedSource,
                    "acquisition-transport-failed");
            }
            finally
            {
                DeleteTree(transaction);
            }
        }
    }

    private static bool IsValid(ProviderArtifactAcquisitionRequest request)
    {
        if (request.Files.Count == 0
            || string.IsNullOrWhiteSpace(request.ExpectedSource.Value)
            || request.MaxTotalBytes <= 0
            || request.Arguments.Any(string.IsNullOrEmpty)
            || !IsSafeRelativePath(request.ExecutablePath)
            || request.Files.Any(file =>
                file is null
                || !IsSafeRelativePath(file.RelativePath)
                || !ContentAddressedProviderStore.IsDigest(file.Sha256)
                || file.Length < 0)
            || request.Files.Select(file => file.RelativePath).Distinct(StringComparer.Ordinal).Count()
                != request.Files.Count
            || !request.Files.Any(file => file.RelativePath == request.ExecutablePath))
        {
            return false;
        }

        long total = 0;
        foreach (var file in request.Files)
        {
            if (file.Length > request.MaxTotalBytes - total)
            {
                return false;
            }

            total += file.Length;
        }

        var identity = ProviderArtifactSetIdentity.Compute(
            request.Files.Select(file => new ProviderArtifactFile(file.RelativePath, file.Sha256)),
            request.ExecutablePath,
            request.Arguments);
        return identity == request.Identity;
    }

    private static bool CopyExact(Stream input, Stream output, long length)
    {
        var remaining = length;
        var buffer = new byte[81920];
        while (remaining > 0)
        {
            var read = input.Read(buffer, 0, (int)Math.Min(buffer.Length, remaining));
            if (read == 0)
            {
                return false;
            }

            output.Write(buffer, 0, read);
            remaining -= read;
        }

        return input.ReadByte() == -1;
    }

    private static bool IsSafeRelativePath(string path)
    {
        if (string.IsNullOrWhiteSpace(path) || Path.IsPathFullyQualified(path) || path.Contains('\\'))
        {
            return false;
        }

        return path.Split('/').All(segment => segment.Length > 0 && segment is not "." and not "..");
    }

    private static string CombineRelative(string root, string relative) =>
        Path.GetFullPath(Path.Combine(root, relative.Replace('/', Path.DirectorySeparatorChar)));

    private static bool IsTransportException(Exception exception) =>
        exception is IOException or UnauthorizedAccessException or ObjectDisposedException or NotSupportedException;

    private static void DeleteTree(string path)
    {
        for (var attempt = 0; ; attempt++)
        {
            if (!Directory.Exists(path))
            {
                return;
            }

            try
            {
                foreach (var file in Directory.EnumerateFiles(path, "*", SearchOption.AllDirectories))
                {
                    File.SetAttributes(file, FileAttributes.Normal);
                }

                Directory.Delete(path, recursive: true);
                return;
            }
            catch (Exception exception) when (
                attempt < 4 && exception is IOException or UnauthorizedAccessException)
            {
                Thread.Sleep(25);
            }
        }
    }
}
